### ISSUE TITLE
ScErrNetworkGatewayTerminated during start of app

### ISSUE THREAD

**Issue by ivarbergman** [Starcounter/Starcounter#3480]
(This issue originates from a discussion with @ivarbergman @per-samuelsson  and @dan31 .)

On one of our production server we experience an error when starting the Launcher app after a `staradmin kill all`:

    staradmin kill all
    star -d=intranet --resourcedir="Launcher\Client" "%cd%\Launcher.exe"

The error output one the console is:

    Launcher -> intranet (starting database ...)
    
    [06:11:49, Starcounter (service)]
    Process exitcode: 2114
    
    [06:11:49, Starcounter (service)]
    ScErrNetworkGatewayTerminated (SCERR10021): The server detected that the network
     gateway executable (scnetworkgateway.exe) unexpectedly terminated.
    

Our solution has been to start the Launcher 2 times, since the 2:d time we start Launcher it works. 

Below is a larger section of the starcounter log

    [06:24:56, Notice, Starcounter (intranet)]
    The database has been unloaded.
    
    [06:24:57, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
    
    [06:25:56, Error, Starcounter (service)]
    ScErrNetworkGatewayTerminated (SCERR10021): The server detected that the network gateway executable (scnetworkgateway.exe) unexpectedly terminated.
    
    [06:25:56, Error, Starcounter (service)]
    Process exitcode: 2114
    
    [06:25:56, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
    
    [06:27:06, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
    
    [06:27:26, Warning, Starcounter.Server (scadminserver)]
    ScErrDbProcNotSignaling (SCERR10027): The database data process, when started, didn't signal it was operational in time. scdata.exe, PID=2756, Database=intranet. Start time: 2016-04-08 06:27:21; time spent waiting: 00:00:05.
    Version: 2.1.1658.
    Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10027.
    
    [06:28:27, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
    
    [06:31:34, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
    
    [06:38:55, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
    
    [06:40:40, Warning, Starcounter.Server (scadminserver)]
    ScErrDbProcNotSignaling (SCERR10027): The database data process, when started, didn't signal it was operational in time. scdata.exe, PID=5056, Database=playground. Start time: 2016-04-08 06:40:35; time spent waiting: 00:00:05.
    Version: 2.1.1658.
    Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR10027.
    
    [06:47:19, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
    
    [06:54:26, Notice, Starcounter (intranet)]
    Checkpoint completed.
    
    [06:56:11, Notice, Starcounter (playground)]
    The database has been unloaded.
    
    [06:56:12, Warning, Starcounter (networkgateway)]
    Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.

**Comment by per-samuelsson** [comment]
Thanks for taking the time to report this @ivarbergman.

The version used was http://downloads.starcounter.com/download/Starcounter/DevelopNightlyBuilds/2.1.1605

**Comment by alemoi** [comment]
Thank you for reporting! Strange that there is no stack trace why gateway process terminates. Also is it reproducible in 100% of the cases?

**Comment by ivarbergman** [comment]
Should have mentioned that on this server we currently use:
    http://downloads.starcounter.com/download/Starcounter/DevelopDailyBuilds/2.1.1658

It used to be reproducible on this specific server before we upgraded to 2.1.1658 (we upgraded from Starcounter-DevelopNightlyBuilds-2.1.1367). However after SC upgrade I don't get this **all** the time.

Just now I've installed a new server, where I get the same error. But this time I don't get an error output in the console. Just this:

![screen shot 2016-04-08 at 14 44 13](https://cloud.githubusercontent.com/assets/112064/14383883/3676a878-fd98-11e5-951d-6068cefa06c0.png)

However in the SC log I get the same message as before.

**Comment by alemoi** [comment]
As from the logs, I can see that gateway is kind of hanging, right?
```
[06:24:57, Warning, Starcounter (networkgateway)]
Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.

[06:25:56, Error, Starcounter (service)]
ScErrNetworkGatewayTerminated (SCERR10021): The server detected that the network gateway executable (scnetworkgateway.exe) unexpectedly terminated.
```

Can you please also change the number of gateway workers to 1 (WorkersNumber scnetworkgateway.xml) and see the effect? Thank you!

**Comment by ivarbergman** [comment]
I'll do some more testning during this week-end, when I can do changes on our servers more freely.

**Comment by alemoi** [comment]
I have produced a daily build for you with some more strict behavior that can possibly reveal the problem: http://downloads.starcounter.com/download/Starcounter/DevelopDailyBuilds/2.1.1687

Please use it @ivarbergman  in your further tests. Thank you!

**Comment by alemoi** [comment]
I made some changes in the code and added more verifications, @ivarbergman can you please test the occurrence of the problem with nightly build #2.1.1761 or later? Thank you!

**Comment by ivarbergman** [comment]
Hi! I have now installed `2.1.1764` on our stage-server. The first restart of the apps after SC update resulted in the same sc log massage as before (no new log messages)
```
2016-04-14 06:19:45	Error	service	Starcounter	Process exitcode: 2114
2016-04-14 06:19:45	Error	service	Starcounter	ScErrNetworkGatewayTerminated (SCERR10021): The server detected that the network gateway executable (scnetworkgateway.exe) unexpectedly terminated.
```

The odd thing after the first restart that failed, I now don't get this error any more.. And this is what also happened on our production-server. After we upgraded to 2.1.1658, the `ScErrNetworkGatewayTerminated` error happened a few times, but after that we don't have this problem any more. Strange..

To continue the test, I reinstalled an older version of SC `2.1.1367` and the corresponding version of our apps. Now the start up error reappeared again. More, if I installed the `2.1.1764` again, the error happened once and after that the app start works fine!

Unfortunate there is no more entries in the sc log, appart from the once we already seen. 

To summarize, newer versions of SC seems to fix the problem with `ScErrNetworkGatewayTerminated`. However it is a bit odd that the first restart of our apps, after SC upgrade, produces the error.
  
It also seems that I can reproduce the error, if I do a installation of an older version of SC.

**Comment by dan31** [comment]
@alemoi @ivarbergman If it happens to appear, most likely the issue is diminished but not resolved. @alemoi, what is your further strategy of fixing? What was the reason for having issue occurring and what does the fix constitute? Maybe if you explain some other people will see and show the best way to resolve it 100%.

**Comment by alemoi** [comment]
The oddest thing for me is that there is no log entry with corresponding reason. That is really unexpected. @ivarbergman thank you for the tests! Did you run with WorkersNumber set to 1?

**Comment by alemoi** [comment]
@ivarbergman can you please give an overview if there is anything special in the `main()` of your startup application? Like WebSockets registration or something specific? Thank you!

**Comment by alemoi** [comment]
I consider issue to be fixed in latest stable/nightly. Please let us know @ivarbergman if you will reproduce the problem again. Thank you!

**Comment by ivarbergman** [comment]
Hi! Sorry for not given you feedback on this issue lately. We will soon
upgrade to the latest stable starcounter, and I will keep an eye on this.

2016-04-20 12:23 GMT+02:00 Alexey <notifications@github.com>:

**Comment by ivarbergman** [comment]
Hi! Today we upgrade our production server to version 2.2.1872 and I got the same error on Luncher start. But this time I see a `Process exitcode: 2114` message in the logs. 

```
20160426T073921 Warning sc://tunityintranet/personal/networkgateway Starcounter - Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
20160426T073944 Debug sc://tunityintranet/personal/intranet Starcounter - Starting recovery at address 106579. Next commit id is 61017.
20160426T074005 Debug sc://tunityintranet/personal/intranet Starcounter - Recovery completed. 3056 transactions recovered.
20160426T074005 Debug sc://tunityintranet/personal/intranet Starcounter - Starting log writer at address 113182. Next commit id is 64073.
20160426T074012 Error sc://tunityintranet/personal/service Starcounter - ScErrNetworkGatewayTerminated (SCERR10021): The server detected that the network gateway executable (scnetworkgateway.exe) unexpectedly terminated.
20160426T074012 Error sc://tunityintranet/personal/service Starcounter - Process exitcode: 2114
20160426T074012 Warning sc://tunityintranet/personal/networkgateway Starcounter - Gateway XML: MaximumReceiveContentLength parameter exceeds maximum value. Current value is set to 2096936.
```

The behaviour is the same as before. The second time I start Launcher it works.

**Comment by alemoi** [comment]
Thank you for reporting! According to the logs that I see that it just exited without any hanging. I will check the code to filter out the possibilities of such silent exit.

**Comment by malx122** [comment]
Any updates about this?

**Comment by alemoi** [comment]
Checking this right now. Did you get again? Under what circumstances?

**Comment by alemoi** [comment]
I added writing of correct error in scservice. If you can try the coming nightly on your machines, that would be great!

**Comment by alemoi** [comment]
@ivarbergman can you please send us your server logs, we just want to check that this error did not occur anymore, or if it occurred we want to see the error code. Did you experience this during recent times? Thanks for the answer!

**Comment by ivarbergman** [comment]
Hi! 
No, we haven't experienced this problem in a while now. So, it seems to be resolved. You can download the logs from our server:

[http://tunitybuild.cloudapp.net:9999/intranet_log_files.zip](http://tunitybuild.cloudapp.net:9999/intranet_log_files.zip)
